src/wp-includes/class-wp-post-type.php

```diff
@@ -670,9 +670,20 @@ public function add_supports() {
 				}
 			}
 			unset( $this->supports );
+
+			/*
+			 * 'editor' support implies 'autosave' support for backward compatibility.
+			 * 'autosave' support needs to be explicitly removed if not desired.
+			 */
+			if (
+				post_type_supports( $this->name, 'editor' ) &&
+				! post_type_supports( $this->name, 'autosave' )
+			) {
+				add_post_type_support( $this->name, 'autosave' );
+			}
```

Comment:
cc @jorgefilipecosta @ellatrix as you've been discussing something similar for footnotes as well.

Reply:
What if the `'autosave_rest_controller_class'` is set to the default of `false`? Could that also imply no `'autosave'` support?

Reply:
It sorta defaults to `false` in `\WP_Post_Type::set_props`, but `\WP_Post_Type::get_autosave_rest_controller()` uses `WP_REST_Autosaves_Controller` if `autosave_rest_controller_class` is falsey`. Hence the workaround in #6034 of using `stdClass`.

That said, even properly changing the default does not help much, as the editor also needs to know about autosave support client-side. With post type supports that's easily doable, see https://github.com/WordPress/wordpress-develop/pull/6035#issuecomment-1929260835

Using post type supports also allows short-circuiting in areas where autosaves are handled, bringing the logic nicely in line with how revisions are a post type support too.

Reply:
While I understand the current BC need, I find this confusing. I'm thinking it'll be hard to document and explain that _only this one specific support works differently_.

Prior to this change, what is configured in a post type's `'supports'` is respected and means: this post only supports these things. 

After this change, this is still true _except for the `'autosave'` support_. It's the only one that is not respected and thus will require explicit removal after registering the post type.

All of Core's post types can be modified in `create_initial_post_types()` to define if each should or shouldn't support `'autosave'`. 

But what about extenders who are registering their post types with specific supports? Yes, they can modify their code to add the `remove_post_type_support()` where needed. But again, it's confusing why they need to do that given the others are respected, but `'autosave'` is not.

NOTE:
This approach may be the only way to ensure BC. I appreciate why it's here. That said, I'm raising my concern to invite further thinking of (a) whether `'autosave'` support should be different and (b) if yes, how best to explain _why_ only this support is different and not respected without doing a  explicit `remove_post_type_support()`.

Reply:
Agreed, it's an outlier. It's the first solution that came to mind to ensure BC.

We could also consider not doing this, as there actually isn't any plugin in the directory overriding `autosave_rest_controller` or `autosave_rest_controller_class`, see https://wpdirectory.net/search/01HNZG1ZZT7PZZKS6S15FF0BQV

Happy to hear other suggestions.

Reply:
>Agreed, it's an outlier. It's the first solution that came to mind to ensure BC.

I can clearly see why it came to mind.

As you noted, if plugins aren't overriding, is there a BC concern? Hmm, maybe not. If there are plugins in the wild (not in the plugins directory) that need to modify their post type `'autosave'` support, that might be better explained in a dev note.

Reply:
Actually, now that I look at it again:

Removing this back compat logic would mean any extender registering a post type with `editor` support would stop supporting autosaves.

Reply:
After sleeping on this and a brief chat with @peterwilsoncc, I'm not seeing a way around the BC concern. Though not a fan of opt-out approaches, here it seems unavoidable.

@peterwilsoncc [pitched an idea to avoid having to use](https://github.com/WordPress/wordpress-develop/pull/6035#issuecomment-1930961395) `remove_post_type_support()` to opt-out after registering. I like the idea, but thinking that approach might be out-of-scope for this ticket and PR. Instead it could be an supports enhancement.